### PR TITLE
feat: add Detray surface grid converter

### PR DESCRIPTION
**Pull request by asalzburger**
This PR adds the surface grid converters drafted by @EleniXoch in #3607 and adds it to the overall `DetrayConverter`.

First tests show that the files written out contain the new surface grids.

Including the surface grids into the navigation reduces the time for the CPU navigation by a factor of  5.5 with identical results.

### CODE REVIEW

File: Plugins/Detray/include/Acts/Plugins/Detray/DetrayConverter.hpp [hunk 1]
```diff
@@ -91,6 +93,19 @@ class DetrayConverter {
       }
     }
 
+    // (3) surface grids
+    if (options.convertSurfaceGrids) {
+      detray::io::detector_grids_payload<std::size_t, detray::io::accel_id>
+          surfaceGridsPayload =
+              DetraySurfaceGridsConverter::convertSurfaceGrids(detector);
+
+      detray::io::surface_grid_reader<typename detector_t::surface_type,
```
Comment (niermann999) [inline Plugins/Detray/include/Acts/Plugins/Detray/DetrayConverter.hpp:102]: Just as an explanation for some of the magic numbers detray requires here
```suggestion
      // Capacity 0 (dynamic bin size) and dimension 2 (2D grids)
      detray::io::surface_grid_reader<typename detector_t::surface_type,
```

File: Plugins/Detray/include/Acts/Plugins/Detray/DetraySurfaceGridsConverter.hpp [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+//
+// Copyright (C) 2024 CERN for the benefit of the Acts project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#pragma once
+
+#include "Acts/Material/MaterialSlab.hpp"
```
Comment (niermann999) [inline Plugins/Detray/include/Acts/Plugins/Detray/DetraySurfaceGridsConverter.hpp:11]:
```suggestion
```

File: Plugins/Detray/include/Acts/Plugins/Detray/DetraySurfaceGridsConverter.hpp [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+#include "Acts/Utilities/TypeList.hpp"
+
+#include <detray/builders/detector_builder.hpp>
+#include <detray/core/detector.hpp>
+#include <detray/definitions/geometry.hpp>
+#include <detray/io/common/geometry_reader.hpp>
+#include <detray/io/common/surface_grid_reader.hpp>
+#include <detray/io/frontend/detector_writer.hpp>
+#include <detray/io/frontend/payloads.hpp>
+#include <detray/utils/consistency_checker.hpp>
```
Comment (niermann999) [inline Plugins/Detray/include/Acts/Plugins/Detray/DetraySurfaceGridsConverter.hpp:25]: Can we check, which of these includes are actually needed?
I would expect:
```suggestion

#include <detray/io/common/surface_grid_reader.hpp>
#include <detray/io/frontend/payloads.hpp>
```

File: Plugins/Detray/src/DetraySurfaceGridsConverter.cpp [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#include "Acts/Plugins/Detray/DetraySurfaceGridsConverter.hpp"
+
+#include "Acts/Detector/Detector.hpp"
+#include "Acts/Material/BinnedSurfaceMaterial.hpp"
+#include "Acts/Material/HomogeneousSurfaceMaterial.hpp"
```
Comment (niermann999) [inline Plugins/Detray/src/DetraySurfaceGridsConverter.cpp]:
```suggestion
```

File: Plugins/Detray/src/DetraySurfaceGridsConverter.cpp [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+template <typename grid_type>
+detray::io::grid_payload<std::size_t, detray::io::accel_id>
+Acts::DetraySurfaceGridsConverter::convertGrid(const grid_type& grid,
+                                               bool swapAxis) {
+  // Get the grid axes & potentially swap them
+  detray::io::grid_payload<std::size_t, detray::io::accel_id> grid_pd;
+
+  auto axes = grid.axes();
+  if (swapAxis && grid_type::DIM == 2u) {
+    // std::cout<<"swap axes"<<std::endl;
```
Comment (niermann999) [inline Plugins/Detray/src/DetraySurfaceGridsConverter.cpp:54]: leftover?

File: Plugins/Detray/src/DetraySurfaceGridsConverter.cpp [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+    }
+  }
+
+  // 2D connections
+  if constexpr (grid_type::DIM == 2u) {
+    for (unsigned int ib0 = 1u; ib0 <= axes[0u]->getNBins(); ++ib0) {
+      for (unsigned int ib1 = 1u; ib1 <= axes[1u]->getNBins(); ++ib1) {
+        typename grid_type::index_t lbin;
+        // Lookup bin - respect swap (if it happened) for the lookup
+        lbin[0u] = swapAxis ? ib1 : ib0;
```
Comment (niermann999) [inline Plugins/Detray/src/DetraySurfaceGridsConverter.cpp:88]: Were these not already swapped in line 54?

File: Plugins/Detray/include/Acts/Plugins/Detray/DetrayConverter.hpp (hunk 1 above)
Comment (asalzburger) [reply Plugins/Detray/include/Acts/Plugins/Detray/DetrayConverter.hpp:102]: Fixed.

File: Plugins/Detray/include/Acts/Plugins/Detray/DetraySurfaceGridsConverter.hpp (hunk 2 above)
Comment (asalzburger) [reply Plugins/Detray/include/Acts/Plugins/Detray/DetraySurfaceGridsConverter.hpp:11]: Removed.

File: Plugins/Detray/include/Acts/Plugins/Detray/DetraySurfaceGridsConverter.hpp (hunk 3 above)
Comment (asalzburger) [reply Plugins/Detray/include/Acts/Plugins/Detray/DetraySurfaceGridsConverter.hpp:25]: Cleaned.

File: Plugins/Detray/src/DetraySurfaceGridsConverter.cpp (hunk 4 above)
Comment (asalzburger) [reply Plugins/Detray/src/DetraySurfaceGridsConverter.cpp]: Fixed.

File: Plugins/Detray/src/DetraySurfaceGridsConverter.cpp (hunk 6 above)
Comment (asalzburger) [reply Plugins/Detray/src/DetraySurfaceGridsConverter.cpp:88]: I think that's correct.

**Review by niermann999 — DISMISSED** [review]

**Comment by asalzburger** [comment]
Hey @niermann999 I had a parameter shadowing warning I removed, can you re-approve, please?

**Review by niermann999 — DISMISSED** [review]

**Comment by asalzburger** [comment]
Failure in vertex finding?

**Comment by asalzburger** [comment]
@niermann999 did you make a new detray version that we can try for fixing this issue ?

**Comment by niermann999** [comment]
> @niermann999 did you make a new detray version that we can try for fixing this issue ?

I fixed all null-derefence warnings in detray, but I still have to get the PR in

File: Plugins/Detray/CMakeLists.txt [hunk 7]
```diff
@@ -5,6 +5,7 @@ add_library(
     src/DetrayConverter.cpp
     src/DetrayGeometryConverter.cpp
     src/DetrayMaterialConverter.cpp
+    src/DetraySurfaceGridsConverter.cpp
 )
 
 add_dependencies(ActsPluginDetray detray::core covfie::core vecmem::core)
```
Comment (niermann999) [inline Plugins/Detray/CMakeLists.txt:11]: ```detray::utils``` does not exist anymore: To get specific detector metadata types (except the default one which is in ```detray::core```) there is ```detray::detectors``` now and to get extended validation tools (probably only for the executables where we do the testing) there is ```detray::test_utils```

File: Plugins/Detray/include/Acts/Plugins/Detray/DetrayGeometryConverter.hpp [hunk 8]
```diff
@@ -17,11 +17,11 @@
 
 #include <vector>
 
+#include <detray/builders/grid_factory.hpp>
 #include <detray/builders/detector_builder.hpp>
 #include <detray/definitions/geometry.hpp>
 #include <detray/io/frontend/detector_writer.hpp>
 #include <detray/io/frontend/payloads.hpp>
```
[the remark this answers is not in the document]
Comment (asalzburger) [reply Plugins/Detray/include/Acts/Plugins/Detray/DetrayGeometryConverter.hpp]: Ok, I will anyway have a follow-up PR where I switch to the new tracers, I think I will tackle it there.

File: Plugins/Detray/CMakeLists.txt (hunk 7 above)
Comment (asalzburger) [reply Plugins/Detray/CMakeLists.txt:11]: Not using `detray::utils` here or?

**Comment by asalzburger** [comment]
Rebased.

**Review by niermann999 — APPROVED** [review]

**Comment by asalzburger** [comment]
> ## [Quality Gate Failed](https://sonarcloud.io/dashboard?id=acts-project_acts&pullRequest=3608) **Quality Gate failed**
> Failed conditions  [32.18% Line Coverage on New Code](https://sonarcloud.io/dashboard?id=acts-project_acts&pullRequest=3608) (required ≥ 50%)
> 
> [See analysis details on SonarCloud](https://sonarcloud.io/dashboard?id=acts-project_acts&pullRequest=3608)

I will overwrite the coverage problem, I will try to add some tests for the last part as well, but it's not really clear to me yet how we will test this in the unit tests.

**Pull request merged**